File: packages/python-client/src/agentvalet/env.py
"""Environment resolution for ``AgentValet.from_env()``.

Ported from ``packages/client/src/env.ts``. Accepts BOTH naming conventions
already in the wild: the bare names the Node CLI and MCPB write
(``AGENT_ID``, ``OWNER_ID``, ``PROXY_URL``, ``AGENT_PRIVATE_KEY*``) and
``AGENTVALET_``-prefixed aliases, which is what the docs tell people to set.

A machine that has run ``agentvalet register`` (or ``npx @agentvalet/register``)
already satisfies the bare set, so ``from_env()`` works with no extra config.
That property is the single most important onboarding affordance in the package
— do not "tidy" it into one canonical name.
"""

from __future__ import annotations

import base64
import os
from pathlib import Path

from ._constants import DEFAULT_PROXY_URL
from .errors import ConfigError

__all__ = ["agent_id_from_env", "owner_id_from_env", "private_key_from_env", "proxy_url_from_env"]

#: Written by the register flow. Also read by the Node SDK and mcp-server.
KEY_PATH = Path.home() / ".agentvalet" / "agent.key"


def _read_env(name: str) -> str | None:
    """Empty strings and unresolved MCPB placeholders count as not-set.

    The MCPB writes a literal ``${user_config.foo}`` when a field is left
    blank. Treating that as a real value produces a baffling 401 much later,
    far from the actual mistake.
    """
    raw = os.environ.get(name)
    if raw is None:
        return None
    value = raw.strip()
    if not value:
        return None
    if value.startswith("${") and value.endswith("}"):
        return None
    return value


def _read_either(suffix: str) -> str | None:
    """``AGENTVALET_``-prefixed name wins over the bare name when both are set."""
    return _read_env(f"AGENTVALET_{suffix}") or _read_env(suffix)


def agent_id_from_env() -> str | None:
    return _read_either("AGENT_ID")


def owner_id_from_env() -> str | None:
    return _read_either("OWNER_ID")


def proxy_url_from_env() -> str:
    return (_read_either("PROXY_URL") or DEFAULT_PROXY_URL).rstrip("/")


def private_key_from_env() -> str:
    """Resolve the RS256 private key PEM.

    Precedence, matching the Node SDK exactly:

    1. ``AGENT_PRIVATE_KEY_B64``  — base64-encoded PEM (12-factor / container safe)
    2. ``AGENT_PRIVATE_KEY_PATH`` — path to a PEM file
    3. ``AGENT_PRIVATE_KEY``      — raw multi-line PEM, or ``\\n``-escaped single line
    4. ``~/.agentvalet/agent.key`` — written by the register flow
    """
    b64 = _read_either("AGENT_PRIVATE_KEY_B64")
    if b64:
        return base64.b64decode(b64).decode("utf-8").strip()

    path = _read_either("AGENT_PRIVATE_KEY_PATH")
    if path:
        try:
            return Path(path).read_text(encoding="utf-8").strip()
        except OSError as err:
            raise ConfigError(f"Cannot read AGENT_PRIVATE_KEY_PATH ({path}): {err}") from err

    raw = _read_either("AGENT_PRIVATE_KEY")
    if raw:
        unescaped = raw.replace("\\n", "\n")
        # Secrets UIs routinely trim the PEM armour on paste. Re-wrap rather
        # than failing deep inside the crypto layer with an opaque message.
        if not unescaped.startswith("-----"):
            return f"-----BEGIN PRIVATE KEY-----\n{unescaped}\n-----END PRIVATE KEY-----"
        return unescaped

    disk = _read_disk_key()
    if disk:
        return disk

    raise ConfigError(
        "No private key found. Set AGENT_PRIVATE_KEY, AGENT_PRIVATE_KEY_PATH, or "
        "AGENT_PRIVATE_KEY_B64, or run `agentvalet register` to create an agent and "
        "write ~/.agentvalet/agent.key."
    )


def _read_disk_key() -> str | None:
    try:
        return KEY_PATH.read_text(encoding="utf-8").strip() or None
    except OSError:
        return None

File: packages/client/src/env.ts
// packages/client/src/env.ts
//
// Environment resolution for AgentValet.fromEnv().
//
// Accepts BOTH naming conventions already in the wild: the bare names the
// mcp-server/MCPB write (AGENT_ID, OWNER_ID, PROXY_URL, AGENT_PRIVATE_KEY*)
// and AGENTVALET_-prefixed aliases, which is what the docs and the
// credentials-not-configured message tell people to set. A machine that has
// been through `npx @agentvalet/register` already satisfies the bare set, so
// fromEnv() works with zero extra config there.

import { readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { ConfigError } from "./errors.js";

export const DEFAULT_PROXY_URL = "https://api.agentvalet.ai";

// Treat empty strings and unresolved MCPB placeholders ("${user_config.foo}")
// as not-set. The MCPB writes literal placeholders when a user_config field is
// left blank, and a literal "${...}" agent id fails far more confusingly later.
function readEnv(name: string): string | null {
  const v = process.env[name];
  if (v === undefined) return null;
  const t = v.trim();
  if (t === "") return null;
  if (t.startsWith("${") && t.endsWith("}")) return null;
  return t;
}

/** AGENTVALET_-prefixed name wins over the bare name when both are set. */
function readEither(suffix: string): string | null {
  return readEnv(`AGENTVALET_${suffix}`) ?? readEnv(suffix);
}

export function agentIdFromEnv(): string | null {
  return readEither("AGENT_ID");
}

export function ownerIdFromEnv(): string | null {
  return readEither("OWNER_ID");
}

export function proxyUrlFromEnv(): string {
  return (readEither("PROXY_URL") ?? DEFAULT_PROXY_URL).replace(/\/$/, "");
}

/**
 * Resolves the RS256 private key PEM, in precedence order:
 *   1. AGENT_PRIVATE_KEY_B64  — base64-encoded PEM (12-factor / container safe)
 *   2. AGENT_PRIVATE_KEY_PATH — path to a PEM file
 *   3. AGENT_PRIVATE_KEY      — raw multi-line PEM, or \n-escaped single line
 *   4. ~/.agentvalet/agent.key — written by `npx @agentvalet/register`
 *
 * Same order and same formats as the mcp-server's pem.ts, so a dev who already
 * has the MCP server working can construct the SDK with no new setup.
 */
export function privateKeyFromEnv(): string {
  const b64 = readEither("AGENT_PRIVATE_KEY_B64");
  if (b64) return Buffer.from(b64, "base64").toString("utf-8").trim();

  const path = readEither("AGENT_PRIVATE_KEY_PATH");
  if (path) {
    try {
      return readFileSync(path, "utf-8").trim();
    } catch (err) {
      throw new ConfigError(
        `Cannot read AGENT_PRIVATE_KEY_PATH (${path}): ${err instanceof Error ? err.message : err}`,
      );
    }
  }

  const raw = readEither("AGENT_PRIVATE_KEY");
  if (raw) {
    const unescaped = raw.replace(/\\n/g, "\n");
    // Tolerate a bare base64 blob with the PEM armour stripped — a common
    // casualty of pasting a key through a secrets UI that trims delimiters.
    if (!unescaped.startsWith("-----")) {
      return `-----BEGIN PRIVATE KEY-----\n${unescaped}\n-----END PRIVATE KEY-----`;
    }
    return unescaped;
  }

  const diskKey = readDiskKey();
  if (diskKey) return diskKey;

  throw new ConfigError(
    "No private key found. Set AGENT_PRIVATE_KEY, AGENT_PRIVATE_KEY_PATH, or AGENT_PRIVATE_KEY_B64, " +
      "or run `npx @agentvalet/register` to create an agent and write ~/.agentvalet/agent.key.",
  );
}

function readDiskKey(): string | null {
  try {
    return readFileSync(join(homedir(), ".agentvalet", "agent.key"), "utf-8").trim() || null;
  } catch {
    return null;
  }
}
